pkg/suggestion/v1alpha3/goptuna/service.go

```diff
@@ -0,0 +1,182 @@
+			//
+			// But suggestion service cannot know which Katib trial name corresponds to Goptuna trial ID.
+			// Because Katib's trial name is determined by Katib controller after finished this gRPC call.
+			// So `findGoptunaTrialIDByParam()` estimates the goptuna trial ID from the parameter values.
+			// See the code comment in `findGoptunaTrialIDByParam()` to understand how estimates Goptuna trial ID.
+			gtrialID, err = findGoptunaTrialIDByParam(s.study, s.trialMapping, ktrial)
+			if err != nil {
+				return err
+			}
+			s.trialMapping[katibTrialName] = gtrialID
+			klog.Infof("Update trial mapping : trialName=%s -> trialID=%d", katibTrialName, gtrialID)
+		}
+
+		gtrial, err := s.study.Storage.GetTrial(gtrialID)
+		if err != nil {
+			return err
+		}
+
+		// It doesn't need to update finished trials.
+		if gtrial.State.IsFinished() {
+			continue
+		}
+
+		if ktrial.State == gtrial.State {
+			continue
+		}
+
+		if ktrial.State == goptuna.TrialStateComplete {
+			err = s.study.Storage.SetTrialValue(gtrialID, ktrial.Value)
+			if err != nil {
+				return err
+			}
+		}
+
+		err = s.study.Storage.SetTrialState(gtrialID, ktrial.State)
+		if err != nil {
+			klog.Errorf("Failed to update state: %s", err)
+			return err
+		}
+
+		klog.Infof("Detect changes of Trial (trialName=%s, trialID=%d) : State %s, Evaluation %f",
+			katibTrialName, gtrialID, ktrial.State, ktrial.Value)
+	}
+	return nil
+}
+
+func (s *SuggestionService) ValidateAlgorithmSettings(
+	ctx context.Context,
+	req *api_v1_alpha3.ValidateAlgorithmSettingsRequest,
+) (*api_v1_alpha3.ValidateAlgorithmSettingsReply, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "request is empty")
+	}
+
+	algorithmName := req.GetExperiment().GetSpec().GetAlgorithm().GetAlgorithmName()
+	if algorithmName != AlgorithmRandom && algorithmName != AlgorithmCMAES && algorithmName != AlgorithmTPE {
+		return nil, status.Error(codes.InvalidArgument, "unsupported algorithm")
+	}
+
+	params := req.GetExperiment().GetSpec().GetParameterSpecs().GetParameters()
+	if algorithmName == AlgorithmCMAES {
+		cnt := 0
+		for _, p := range params {
+			if p.ParameterType == api_v1_alpha3.ParameterType_DOUBLE || p.ParameterType == api_v1_alpha3.ParameterType_INT {
+				cnt++
+			}
+		}
+		if cnt < 2 {
+			return nil, status.Error(codes.InvalidArgument, "CMA-ES only supports two or more dimensional continuous search space.")
+		}
+	}
+
+	paramSet := make(map[string]interface{}, len(params))
+	for _, p := range params {
+		if _, ok := paramSet[p.Name]; ok {
+			return nil, status.Errorf(codes.InvalidArgument, "Detect duplicated parameter name: %s", p.Name)
+		}
+		paramSet[p.Name] = nil
+	}
+
+	study, searchSpace, err := createStudyAndSearchSpace(req.GetExperiment())
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to create goptuna study: %s", err.Error())
+	}
+	klog.Infof("Goptuna search space: %#v", searchSpace)
+
+	s.mu.Lock()
```

Comment:
Why do you need mutex here?

Reply:
If received `ValidateAlgorithmSettings` and `GetSuggestions` requests at almost the same time, following scenario might happen.

1. In the request of `ValidateAlgorithmSettings`, `s.study` is set at L175.
1. In the request of `GetSuggestions`, `if s.study == nil || s.searchSpace == nil {}` (L41-L43) is true even if `s.searchSpace` is nil.

Reply:
Ah, I forgot to set RLock for L41-L43.

Reply:
Do avoid this situation, why you don't want to create `s.study` and `s.searchSpace` in the first GetSuggestion call?
You just need 1 additional variable. In that case, you don't need mutex, etc. 
Like we did here: https://github.com/kubeflow/katib/blob/master/pkg/suggestion/v1alpha3/hyperopt_service.py#L28.

I think in `ValidateAlgorithmSettings` you just need to validate Search Space and Algorithm Settings.
What do you think?

Reply:
Although `hyperopt` suggestion service works on multiple threads like [here](https://github.com/kubeflow/katib/blob/859f665c278913c1f5f14a8b5ffcb087505f3116/cmd/suggestion/hyperopt/v1alpha3/main.py#L13), it looks that the current implementation of `hyperopt` service is not thread-safe. 

The first `GetSuggestions()` request is received, `self.is_first_run` is True. But if received second `GetSuggestions` request while the first request still executes the following lines:

https://github.com/kubeflow/katib/blob/859f665c278913c1f5f14a8b5ffcb087505f3116/pkg/suggestion/v1alpha3/hyperopt_service.py#L29-L33

Is it ensured that katib-controller doesn't send concurrent grpc requests to suggestion service?

Reply:
I believe, for the first `GetSuggestion` call, Katib controller will not send  concurrent grpc requests.

Only the case, when Katib Controller can call `GetSuggestion` again, if timeout (https://github.com/kubeflow/katib/blob/master/pkg/controller.v1alpha3/suggestion/suggestionclient/suggestionclient.go#L66) is over, which currently is 1 minute.

Reply:
I see. Do you think it's better to remove the goroutine lock like `hyperopt` service?

Reply:
If you are worried about concurrent calls maybe add `mutex` to `GetSuggestions`?
Since only there we can get concurrent calls.
Also, I think, you should move 
`s.study = study
s.searchSpace = searchSpace` to the `GetSuggestions` call, as I mentioned above.

Reply:
I see 👍 I keep to use `mutex` lock because it has no disadvantages in terms of performance.

I pushed a9be4050, study and search space is created at the first `GetSuggestoins` request.